Added User entity

slizko1 opened the pull request:
Added the required DTOs and mapper
Added AuthenticationController
Added FieldMatch annotation
Added handling the RegistrationException in GlobalExceptionHandler

slizko1 committed:
Added User entity
Added the required DTOs and mapper
Added AuthenticationController
Added FieldMatch annotation
Added handling the RegistrationException in GlobalExceptionHandler

karma-o commented on src/main/java/com/samoilenko/onlinebookstore/controller/AuthentificationController.java:
```diff
@@ -0,0 +1,25 @@
+    private final UserService userService;
+
+    @PostMapping("/registration")
+    public UserResponseDto register(@RequestBody UserRegistrationRequestDto request)
```
did you test your validation? because it does not work without this:
```suggestion
    public UserResponseDto register(@RequestBody @Valid UserRegistrationRequestDto request)
```

karma-o commented on src/main/java/com/samoilenko/onlinebookstore/exception/CustomGlobalExceptionHandler.java:
```diff
@@ -39,6 +39,14 @@ protected ResponseEntity<Object> handleEntityNotFoundException(
+            RegistrationException ex, WebRequest request
+    ) {
+        String bodyOfResponse = "Can't register user" + ex.getMessage();
+        return getResponseEntity(HttpStatus.CONFLICT, bodyOfResponse);
```
https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/409

> The HTTP 409 Conflict response status code indicates a request conflict with the current state of the target resource.
Conflicts are most likely to occur in response to a [PUT](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/PUT) request. For example, you may get a 409 response when uploading a file that is older than the existing one on the server, resulting in a version control conflict.

Registration is about the POST request and creating the entity, to me this looks more like a usual `400 - bad request`

karma-o commented on src/main/java/com/samoilenko/onlinebookstore/exception/CustomGlobalExceptionHandler.java:
```diff
@@ -39,6 +39,14 @@ protected ResponseEntity<Object> handleEntityNotFoundException(
+    protected ResponseEntity<Object> handleRegistrationException(
+            RegistrationException ex, WebRequest request
+    ) {
+        String bodyOfResponse = "Can't register user" + ex.getMessage();
```
`body` is confusing because I thought that it is the whole json body, when in fact it is just a message
```suggestion
        String errorMessage = "Can't register user" + ex.getMessage();
```

karma-o commented on src/main/java/com/samoilenko/onlinebookstore/mapper/UserMapper.java:
```diff
@@ -0,0 +1,21 @@
+@Mapper(config = MapperConfig.class)
+public interface UserMapper {
+
+    UserDto toDto(User user);
+
+    UserResponseDto toResponseDto(User user);
```
no need to have 2 methods here, because we do not need the `UserDto` with a password in it, let's leave only `UserResponseDto` and a method for it

karma-o commented on src/main/java/com/samoilenko/onlinebookstore/service/impl/UserServiceImpl.java:
```diff
@@ -0,0 +1,27 @@
+
+    public UserResponseDto register(UserRegistrationRequestDto requestDto)
+            throws RegistrationException {
+        if (userRepository.findByEmail(requestDto.getEmail()).isPresent()) {
```
can be a bit faster than fetching the data that we do not use
```suggestion
        if (userRepository.existsByEmail(requestDto.getEmail())) {
```

karma-o commented on src/main/java/com/samoilenko/onlinebookstore/validation/constraints/FieldMatch.java:
```diff
@@ -0,0 +1,31 @@
+    String first();
+    String second();
+
+    @Target({TYPE, ANNOTATION_TYPE})
+    @Retention(RUNTIME)
+    @Documented
+    @interface List {
+        FieldMatch[] value();
+    }
```
not sure we need this

karma-o commented on src/main/java/com/samoilenko/onlinebookstore/validation/constraints/FieldMatchValidator.java:
```diff
@@ -0,0 +1,19 @@
+import jakarta.validation.ConstraintValidator;
+import jakarta.validation.ConstraintValidatorContext;
+
+public class FieldMatchValidator implements
+        ConstraintValidator<FieldMatch, UserRegistrationRequestDto> {
```
* first and second fields that you pass to the annotation are not used here
* Annotation is Called `FieldMatch` not a **password** match - we should validate if 2 fields that we received as an arguments in annotation are equal
* use reflection to get the value of those 2 fields
* And it should be applicable not only for `UserRegistrationRequestDto` but for any class
```suggestion
public class FieldMatchValidator implements
        ConstraintValidator<FieldMatch, Object> {
```

karma-o commented on src/main/resources/db/changelog/changes/02-create-users-table.yaml:
```diff
@@ -0,0 +1,38 @@
+                    nullable: false
+              - column:
+                  name: shipping_Address
+                  type: VARCHAR(500)
```
```suggestion
                  type: VARCHAR(500)
                  
```

karma-o commented on src/main/resources/db/changelog/db.changelog-master.yaml:
```diff
@@ -1,3 +1,5 @@
   - include:
       file: db/changelog/changes/01-create-books-table.yaml
+  - include:
+      file: db/changelog/changes/02-create-users-table.yaml
```
```suggestion
      file: db/changelog/changes/02-create-users-table.yaml

```

karma-o reviewed (changes requested):
So far so good! Need to make several improvements

slizko1 committed:
Some fixes were implemented

slizko1 committed:
Added spring-boot-starter-security.
Implemented SecurityConfig with PasswordEncoder and SecurityFilterChain beans.
Implemented UserDetails interface in the User class.
Added implementation for UserDetailsService.
Introduced Role entity. Created RoleRepository.
Added Set<Role> roles field to the User entity.
Added Liquibase changesets to insert roles into the database and assign roles for existing users.

karma-o commented on src/main/java/com/samoilenko/onlinebookstore/validation/constraints/FieldMatchValidator.java:
```diff
@@ -0,0 +1,37 @@
+            Object firstValue = firstField.get(fieldsSource);
+            Object secondValue = secondField.get(fieldsSource);
+
+            return firstValue == null && secondValue == null
+                    || firstValue != null && firstValue.equals(secondValue);
```
Use `Objects.equals()` instead

karma-o commented on src/main/java/com/samoilenko/onlinebookstore/validation/constraints/FieldMatchValidator.java:
```diff
@@ -0,0 +1,37 @@
+            return firstValue == null && secondValue == null
+                    || firstValue != null && firstValue.equals(secondValue);
+        } catch (Exception e) {
+            return false;
```
I'd rather re-throw some exception here instead of returning false because it is our fault as a programmer if we pass an invalid field to the annotation parameter or won't be able to access it

karma-o commented on src/main/resources/db/changelog/changes/02-create-users-table.yaml:
```diff
@@ -0,0 +1,38 @@
+                  constraints:
+                    nullable: false
+              - column:
+                  name: first_Name
```
same for other fields
```suggestion
                  name: first_name
```

karma-o reviewed (approved):
GJ!

slizko1 committed:
New exception and some fixes was added
